src/convnets/resnets/core.jl

```diff
@@ -0,0 +1,468 @@
+
+"""
+    basicblock(inplanes, planes; stride = 1, downsample = identity,
+               reduction_factor = 1, dilation = 1, first_dilation = dilation,
+               activation = relu, connection = addact\$activation,
+               norm_layer = BatchNorm, drop_block = identity, drop_path = identity,
+               attn_fn = planes -> identity, attn_args::NamedTuple = NamedTuple())
+
+Creates a basic ResNet block.
+
+# Arguments
+
+  - `inplanes`: number of input feature maps
+  - `planes`: number of feature maps for the block
+  - `stride`: the stride of the block
+  - `downsample`: the downsampling function to use
+  - `reduction_factor`: the reduction factor that the input feature maps are reduced by before the first
+    convolution.
+  - `dilation`: the dilation of the second convolution.
+  - `first_dilation`: the dilation of the first convolution.
+  - `activation`: the activation function to use.
+  - `connection`: the function applied to the output of residual and skip paths in
+    a block. See [`addact`](#) and [`actadd`](#) for an example. Note that this uses
+    PartialFunctions.jl to pass in the activation function with the notation `addact\$activation`.
+  - `norm_layer`: the normalization layer to use.
+  - `drop_block`: the drop block layer. This is usually initialised in the `_make_blocks`
+    function and passed in.
+  - `drop_path`: the drop path layer. This is usually initialised in the `_make_blocks`
+    function and passed in.
+  - `attn_fn`: the attention function to use. See [`squeeze_excite`](#) for an example.
+  - `attn_args`: a NamedTuple that contains none, some or all of the arguments to be passed to the
+    attention function.
+"""
+function basicblock(inplanes, planes; stride = 1, downsample = identity,
+                    reduction_factor = 1, dilation = 1, first_dilation = dilation,
+                    activation = relu, connection = addact$activation,
+                    norm_layer = BatchNorm, drop_block = identity, drop_path = identity,
+                    attn_fn = planes -> identity, attn_args::NamedTuple = NamedTuple())
+    expansion = expansion_factor(basicblock)
+    first_planes = planes ÷ reduction_factor
+    outplanes = planes * expansion
+    conv_bn1 = Chain(Conv((3, 3), inplanes => first_planes; stride, pad = first_dilation,
+                          dilation = first_dilation, bias = false),
+                     norm_layer(first_planes))
+    drop_block = drop_block
+    conv_bn2 = Chain(Conv((3, 3), first_planes => outplanes; pad = dilation,
+                          dilation = dilation, bias = false),
+                     norm_layer(outplanes))
+    attn_layer = attn_fn(outplanes; attn_args...)
+    return Parallel(connection, downsample,
+                    Chain(conv_bn1, drop_block, activation, conv_bn2, attn_layer,
+                          drop_path))
+end
+expansion_factor(::typeof(basicblock)) = 1
```

Comment:
We still need this to make the user's life easier. Making it a keyword means that we need an extra assert to make sure that someone doesn't play around with this, because it's not meant to be changed. And hardcoding it is impossible because expansion is used outside of the block - it's the only argument which is, in fact